From e2c48c0b711696cfe0381ab53f4366711e7b285d Mon Sep 17 00:00:00 2001
From: Jagpal Singh Gill <paligill@gmail.com>
Date: Tue, 29 Jul 2025 21:01:23 -0700
Subject: [PATCH] Implement valve monitor service

The purpose of the valve monitor service is to monitor the water valves
in liquid cooling setups and raise an event if the valve open/close
status changes. The valve open status has been exposed as a sensor as
Redfish has added Valve as a reading under sensors schema and
CoolingConnector also represents Valve as a sensor. More details on the
implementation are covered in Readme.md.

Related PRs:
Add Valve as inventory item [1]
phosphor-dbus-interface changes for valve inventory item [2]
EM Configuration schema [3]
EM Configuration phosphor-dbus-interfaces [4]
Updates to Sensor.Value [5]
Valve Event interface [6]
Valve control interface [7]

[1] https://gerrit.openbmc.org/c/openbmc/entity-manager/+/77415
[2] https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/77416
[3] https://gerrit.openbmc.org/c/openbmc/entity-manager/+/82310
[4] https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/82312
[5] https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/82304
[6] https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/77419
[7] https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/82527

Tested:
Tested using gsim on a qemu
```
`- /xyz
  `- /xyz/openbmc_project
    |- /xyz/openbmc_project/control
    | `- /xyz/openbmc_project/control/valve_open
    |   |- /xyz/openbmc_project/control/valve_open/ReturnValve
    |   `- /xyz/openbmc_project/control/valve_open/SupplyValve
    `- /xyz/openbmc_project/sensors
      `- /xyz/openbmc_project/sensors/valve_open
        |- /xyz/openbmc_project/sensors/valve_open/ReturnValve
        `- /xyz/openbmc_project/sensors/valve_open/SupplyValve

NAME                                                  TYPE      SIGNATURE RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable                   interface -         -                                        -
.Introspect                                           method    -         s                                        -
org.freedesktop.DBus.Peer                             interface -         -                                        -
.GetMachineId                                         method    -         s                                        -
.Ping                                                 method    -         -                                        -
org.freedesktop.DBus.Properties                       interface -         -                                        -
.Get                                                  method    ss        v                                        -
.GetAll                                               method    s         a{sv}                                    -
.Set                                                  method    ssv       -                                        -
.PropertiesChanged                                    signal    sa{sv}as  -                                        -
xyz.openbmc_project.Association.Definitions           interface -         -                                        -
.Associations                                         property  a(sss)    2 "inventory" "sensors" "/xyz/openbmc... emits-change writable
xyz.openbmc_project.Sensor.Value                      interface -         -                                        -
.MaxValue                                             property  d         100                                      emits-change writable
.MinValue                                             property  d         0                                        emits-change writable
.Unit                                                 property  s         "xyz.openbmc_project.Sensor.Value.Uni... emits-change writable
.Value                                                property  d         100                                      emits-change writable
xyz.openbmc_project.State.Decorator.Availability      interface -         -                                        -
.Available                                            property  b         true                                     emits-change writable
xyz.openbmc_project.State.Decorator.OperationalStatus interface -         -                                        -
.Functional                                           property  b         true                                     emits-change writable

// Toggle the gpio to make value change affect

NAME                                                  TYPE      SIGNATURE RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable                   interface -         -                                        -
.Introspect                                           method    -         s                                        -
org.freedesktop.DBus.Peer                             interface -         -                                        -
.GetMachineId                                         method    -         s                                        -
.Ping                                                 method    -         -                                        -
org.freedesktop.DBus.Properties                       interface -         -                                        -
.Get                                                  method    ss        v                                        -
.GetAll                                               method    s         a{sv}                                    -
.Set                                                  method    ssv       -                                        -
.PropertiesChanged                                    signal    sa{sv}as  -                                        -
xyz.openbmc_project.Association.Definitions           interface -         -                                        -
.Associations                                         property  a(sss)    2 "inventory" "sensors" "/xyz/openbmc... emits-change writable
xyz.openbmc_project.Sensor.Value                      interface -         -                                        -
.MaxValue                                             property  d         100                                      emits-change writable
.MinValue                                             property  d         0                                        emits-change writable
.Unit                                                 property  s         "xyz.openbmc_project.Sensor.Value.Uni... emits-change writable
.Value                                                property  d         0                                        emits-change writable
xyz.openbmc_project.State.Decorator.Availability      interface -         -                                        -
.Available                                            property  b         true                                     emits-change writable
xyz.openbmc_project.State.Decorator.OperationalStatus interface -         -                                        -
.Functional                                           property  b         true                                     emits-change writable

...
Apr 03 13:49:55 ventura valvemonitor[4371]: OPENBMC_MESSAGE_ID={"xyz.openbmc_project.State.Valve.ValveClose":{"VALVE_NAME":"/xyz/openbmc_project/inventory/system/valve/Ventura_Supply_Valve","_SOURCE":{"COLUMN":76,"FILE":"../git/src/valve-monitor/ValveEvents.cpp","FUNCTION":"sdbusplus::async::task<> valve::Events::generateValveEvent(sdbusplus::object_path, bool)","LINE":21,"PID":4371}}}
...

```

Change-Id: I1ccec2b43d02605e8f2e7df90fb534174eecc34e
Signed-off-by: Jagpal Singh Gill <paligill@gmail.com>
---
 service_files/meson.build                     |   1 +
 .../xyz.openbmc_project.valvemonitor.service  |  15 +
 src/{GPIOInterface.cpp => GPIInterface.cpp}   |  40 +--
 src/GPIInterface.hpp                          |  44 +++
 src/GPIOInterface.hpp                         |  42 ---
 src/GPOInterface.cpp                          |  51 +++
 src/GPOInterface.hpp                          |  28 ++
 src/leakdetector/LeakGPIODetector.cpp         |   8 +-
 src/leakdetector/LeakGPIODetector.hpp         |   4 +-
 src/leakdetector/meson.build                  |   2 +-
 src/meson.build                               |  22 +-
 src/valve-monitor/GPIOValve.cpp               | 299 ++++++++++++++++++
 src/valve-monitor/GPIOValve.hpp               | 105 ++++++
 src/valve-monitor/Readme.md                   |  50 +++
 src/valve-monitor/ValveEvents.cpp             |  34 ++
 src/valve-monitor/ValveEvents.hpp             |  23 ++
 src/valve-monitor/ValveMonitor.cpp            | 102 ++++++
 src/valve-monitor/ValveMonitor.hpp            |  49 +++
 src/valve-monitor/meson.build                 |  22 ++
 19 files changed, 862 insertions(+), 79 deletions(-)
 create mode 100644 service_files/xyz.openbmc_project.valvemonitor.service
 rename src/{GPIOInterface.cpp => GPIInterface.cpp} (55%)
 create mode 100644 src/GPIInterface.hpp
 delete mode 100644 src/GPIOInterface.hpp
 create mode 100644 src/GPOInterface.cpp
 create mode 100644 src/GPOInterface.hpp
 create mode 100644 src/valve-monitor/GPIOValve.cpp
 create mode 100644 src/valve-monitor/GPIOValve.hpp
 create mode 100644 src/valve-monitor/Readme.md
 create mode 100644 src/valve-monitor/ValveEvents.cpp
 create mode 100644 src/valve-monitor/ValveEvents.hpp
 create mode 100644 src/valve-monitor/ValveMonitor.cpp
 create mode 100644 src/valve-monitor/ValveMonitor.hpp

diff --git a/service_files/meson.build b/service_files/meson.build
index 5b949a2..4a7c776 100644
--- a/service_files/meson.build
+++ b/service_files/meson.build
@@ -21,6 +21,7 @@ unit_files = [
     ['nvme', 'xyz.openbmc_project.nvmesensor.service'],
     ['psu', 'xyz.openbmc_project.psusensor.service'],
     ['smbpbi', 'xyz.openbmc_project.smbpbisensor.service'],
+    ['valve-monitor', 'xyz.openbmc_project.valvemonitor.service'],
 ]
 
 fs = import('fs')
diff --git a/service_files/xyz.openbmc_project.valvemonitor.service b/service_files/xyz.openbmc_project.valvemonitor.service
new file mode 100644
index 0000000..713e372
--- /dev/null
+++ b/service_files/xyz.openbmc_project.valvemonitor.service
@@ -0,0 +1,15 @@
+
+[Unit]
+Description=Valve Monitor for Liquid Cooling
+Requires=xyz.openbmc_project.EntityManager.service
+After=xyz.openbmc_project.EntityManager.service
+
+[Service]
+Type=dbus
+BusName=xyz.openbmc_project.valvemonitor
+Restart=always
+RestartSec=5
+ExecStart=/usr/libexec/dbus-sensors/valvemonitor
+
+[Install]
+WantedBy=multi-user.target
diff --git a/src/GPIOInterface.cpp b/src/GPIInterface.cpp
similarity index 55%
rename from src/GPIOInterface.cpp
rename to src/GPIInterface.cpp
index 81a93d3..b023c37 100644
--- a/src/GPIOInterface.cpp
+++ b/src/GPIInterface.cpp
@@ -1,10 +1,9 @@
-#include "GPIOInterface.hpp"
+#include "GPIInterface.hpp"
 
 #include <gpiod.hpp>
 #include <phosphor-logging/lg2.hpp>
 #include <sdbusplus/async.hpp>
 
-#include <chrono>
 #include <exception>
 #include <memory>
 #include <stdexcept>
@@ -16,10 +15,10 @@ namespace gpio
 
 PHOSPHOR_LOG2_USING;
 
-GPIOInterface::GPIOInterface(sdbusplus::async::context& ctx,
-                             const std::string& consumerName,
-                             const std::string& pinName, bool activeLow,
-                             Callback_t updateStateCallback) :
+GPIInterface::GPIInterface(sdbusplus::async::context& ctx,
+                           const std::string& consumerName,
+                           const std::string& pinName, bool activeLow,
+                           Callback_t updateStateCallback) :
     ctx(ctx), pinName(pinName),
     updateStateCallback(std::move(updateStateCallback))
 {
@@ -30,7 +29,7 @@ GPIOInterface::GPIOInterface(sdbusplus::async::context& ctx,
     line = gpiod::find_line(pinName);
     if (!line)
     {
-        throw std::runtime_error("Failed to find GPIO line for " + pinName);
+        throw std::runtime_error("Failed to find GPI line for " + pinName);
     }
     try
     {
@@ -47,27 +46,27 @@ GPIOInterface::GPIOInterface(sdbusplus::async::context& ctx,
     if (lineFd < 0)
     {
         throw std::runtime_error(
-            "Failed to get event fd for GPIO line " + pinName);
+            "Failed to get event fd for GPI line " + pinName);
     }
 
     fdioInstance = std::make_unique<sdbusplus::async::fdio>(ctx, lineFd);
 }
 
-auto GPIOInterface::start() -> sdbusplus::async::task<>
+auto GPIInterface::start() -> sdbusplus::async::task<>
 {
-    // Start the async read for the GPIO line
-    ctx.spawn(readGPIOAsyncEvent());
+    // Start the async read for the GPI line
+    ctx.spawn(readGPIAsyncEvent());
 
-    // Read the initial GPIO value
-    co_await readGPIOAsync();
+    // Read the initial GPI value
+    co_await readGPIAsync();
 }
 
-auto GPIOInterface::readGPIOAsync() -> sdbusplus::async::task<>
+auto GPIInterface::readGPIAsync() -> sdbusplus::async::task<>
 {
     auto lineValue = line.get_value();
     if (lineValue < 0)
     {
-        error("Failed to read GPIO line {LINENAME}", "LINENAME", pinName);
+        error("Failed to read GPI line {LINENAME}", "LINENAME", pinName);
         co_return;
     }
     co_await updateStateCallback(lineValue == gpiod::line_event::RISING_EDGE);
@@ -75,22 +74,13 @@ auto GPIOInterface::readGPIOAsync() -> sdbusplus::async::task<>
     co_return;
 }
 
-auto GPIOInterface::readGPIOAsyncEvent() -> sdbusplus::async::task<>
+auto GPIInterface::readGPIAsyncEvent() -> sdbusplus::async::task<>
 {
     while (!ctx.stop_requested())
     {
         // Wait for the fd event for the line to change
         co_await fdioInstance->next();
 
-        // event_read() does not clear the EPOLLIN flag immediately; it is
-        // cleared during the subsequent epoll check. Therefore, call event_wait
-        // with a zero timeout to ensure that event_read() is only invoked when
-        // an event is available, preventing it from blocking.
-        if (!line.event_wait(std::chrono::milliseconds(0)))
-        {
-            continue;
-        }
-
         line.event_read();
         auto lineValue = line.get_value();
 
diff --git a/src/GPIInterface.hpp b/src/GPIInterface.hpp
new file mode 100644
index 0000000..c12f6b6
--- /dev/null
+++ b/src/GPIInterface.hpp
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <gpiod.hpp>
+#include <sdbusplus/async.hpp>
+
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace gpio
+{
+
+/** @brief Class to handle General Purpose Input interface */
+class GPIInterface
+{
+  public:
+    using Callback_t = std::function<sdbusplus::async::task<>(bool)>;
+
+    GPIInterface() = delete;
+
+    /** @brief Constructor to configure the GPI interface in read event mode */
+    GPIInterface(sdbusplus::async::context& ctx,
+                 const std::string& consumerName, const std::string& pinName,
+                 bool activeLow, Callback_t updateStateCallback);
+
+    /** @brief Start the GPI Interface */
+    auto start() -> sdbusplus::async::task<>;
+
+  private:
+    /** @brief Read the gpi state asynchronously */
+    auto readGPIAsync() -> sdbusplus::async::task<>;
+
+    /** @brief Read the gpi state asynchronously based on gpi event */
+    auto readGPIAsyncEvent() -> sdbusplus::async::task<>;
+
+    sdbusplus::async::context& ctx;
+    const std::string& pinName;
+    Callback_t updateStateCallback;
+    gpiod::line line;
+    /** File descriptor based async event handler */
+    std::unique_ptr<sdbusplus::async::fdio> fdioInstance;
+};
+
+} // namespace gpio
diff --git a/src/GPIOInterface.hpp b/src/GPIOInterface.hpp
deleted file mode 100644
index b860551..0000000
--- a/src/GPIOInterface.hpp
+++ /dev/null
@@ -1,42 +0,0 @@
-#pragma once
-
-#include <gpiod.hpp>
-#include <sdbusplus/async.hpp>
-
-#include <functional>
-#include <memory>
-#include <string>
-
-namespace gpio
-{
-
-class GPIOInterface
-{
-  public:
-    using Callback_t = std::function<sdbusplus::async::task<>(bool)>;
-
-    GPIOInterface() = delete;
-
-    GPIOInterface(sdbusplus::async::context& ctx,
-                  const std::string& consumerName, const std::string& pinName,
-                  bool activeLow, Callback_t updateStateCallback);
-
-    /** @brief Start the GPIO Interface */
-    auto start() -> sdbusplus::async::task<>;
-
-  private:
-    /** @brief Read the gpio state asynchronously */
-    auto readGPIOAsync() -> sdbusplus::async::task<>;
-
-    /** @brief Read the gpio state asynchronously based on gpio event */
-    auto readGPIOAsyncEvent() -> sdbusplus::async::task<>;
-
-    sdbusplus::async::context& ctx;
-    const std::string& pinName;
-    Callback_t updateStateCallback;
-    gpiod::line line;
-    /** File descriptor based async event handler */
-    std::unique_ptr<sdbusplus::async::fdio> fdioInstance;
-};
-
-} // namespace gpio
diff --git a/src/GPOInterface.cpp b/src/GPOInterface.cpp
new file mode 100644
index 0000000..ed61f5c
--- /dev/null
+++ b/src/GPOInterface.cpp
@@ -0,0 +1,51 @@
+#include "GPOInterface.hpp"
+
+#include <gpiod.hpp>
+#include <phosphor-logging/lg2.hpp>
+
+#include <exception>
+#include <stdexcept>
+#include <string>
+
+namespace gpio
+{
+
+PHOSPHOR_LOG2_USING;
+
+GPOInterface::GPOInterface(const std::string& consumerName,
+                           const std::string& pinName) : pinName(pinName)
+{
+    line = gpiod::find_line(pinName);
+    if (!line)
+    {
+        throw std::runtime_error("Failed to find output line for " + pinName);
+    }
+    try
+    {
+        line.request({consumerName, gpiod::line_request::DIRECTION_OUTPUT, 0});
+    }
+    catch (std::exception& e)
+    {
+        throw std::runtime_error("Failed to request line for " + pinName +
+                                 " with error " + e.what());
+    }
+}
+
+auto GPOInterface::setValue(bool value) -> bool
+{
+    try
+    {
+        line.set_value(value ? 1 : 0);
+    }
+    catch (std::exception& e)
+    {
+        error(
+            "Failed to set value for output line {LINENAME} with error {ERROR}",
+            "LINENAME", pinName, "ERROR", e);
+        return false;
+    }
+
+    return true;
+}
+
+} // namespace gpio
diff --git a/src/GPOInterface.hpp b/src/GPOInterface.hpp
new file mode 100644
index 0000000..22ff9ed
--- /dev/null
+++ b/src/GPOInterface.hpp
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <gpiod.hpp>
+
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace gpio
+{
+
+/** @brief Class to handle General Purpose Output interface */
+class GPOInterface
+{
+  public:
+    GPOInterface() = delete;
+
+    GPOInterface(const std::string& consumerName, const std::string& pinName);
+
+    /** @brief Set the GPO value */
+    auto setValue(bool value) -> bool;
+
+  private:
+    const std::string& pinName;
+    gpiod::line line;
+};
+
+} // namespace gpio
diff --git a/src/leakdetector/LeakGPIODetector.cpp b/src/leakdetector/LeakGPIODetector.cpp
index b8a2012..ed0b330 100644
--- a/src/leakdetector/LeakGPIODetector.cpp
+++ b/src/leakdetector/LeakGPIODetector.cpp
@@ -48,15 +48,15 @@ GPIODetector::GPIODetector(sdbusplus::async::context& ctx, Events& leakEvents,
                  DetectorIntf::Detector::properties_t{
                      config.name, DetectorState::Normal, config.type}),
     ctx(ctx), leakEvents(leakEvents), config(config),
-    gpioInterface(ctx, config.name, config.pinName,
-                  (config.polarity == config::PinPolarity::activeLow),
-                  std::bind_front(&GPIODetector::updateGPIOStateAsync, this))
+    inputInterface(ctx, config.name, config.pinName,
+                   (config.polarity == config::PinPolarity::activeLow),
+                   std::bind_front(&GPIODetector::updateGPIOStateAsync, this))
 {
     Detector::emit_added();
 
     createAssociations();
 
-    ctx.spawn(gpioInterface.start());
+    ctx.spawn(inputInterface.start());
 
     debug("Created leak detector {NAME}", "NAME", config.name);
 }
diff --git a/src/leakdetector/LeakGPIODetector.hpp b/src/leakdetector/LeakGPIODetector.hpp
index 8c27f0e..0e22232 100644
--- a/src/leakdetector/LeakGPIODetector.hpp
+++ b/src/leakdetector/LeakGPIODetector.hpp
@@ -1,6 +1,6 @@
 #pragma once
 
-#include "GPIOInterface.hpp"
+#include "GPIInterface.hpp"
 #include "LeakEvents.hpp"
 
 #include <phosphor-logging/lg2.hpp>
@@ -100,7 +100,7 @@ class GPIODetector : public DetectorIntf
     sdbusplus::async::context& ctx;
     Events& leakEvents;
     config::DetectorConfig config;
-    gpio::GPIOInterface gpioInterface;
+    gpio::GPIInterface inputInterface;
 };
 
 } // namespace leak
diff --git a/src/leakdetector/meson.build b/src/leakdetector/meson.build
index d2fdfaa..4d0fe04 100644
--- a/src/leakdetector/meson.build
+++ b/src/leakdetector/meson.build
@@ -14,7 +14,7 @@ executable(
         utils_dep,
         phosphor_dbus_interfaces_dep,
         gpiodcxx,
-        gpiointerface_dep,
+        gpiinterface_dep,
         entitymanagerinterface_dep,
     ],
     include_directories: src_inc,
diff --git a/src/meson.build b/src/meson.build
index 3ecf709..2e7d82e 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -63,15 +63,27 @@ pwmsensor_dep = declare_dependency(
 
 systemdinterface_src = files('SystemdInterface.cpp')
 
-gpiointerface_a = static_library(
-    'gpiointerface_a',
-    'GPIOInterface.cpp',
+gpiinterface_a = static_library(
+    'gpiinterface_a',
+    'GPIInterface.cpp',
     dependencies: [default_deps, gpiodcxx],
 )
 
-gpiointerface_dep = declare_dependency(
+gpiinterface_dep = declare_dependency(
     include_directories: ['.'],
-    link_with: [gpiointerface_a],
+    link_with: [gpiinterface_a],
+    dependencies: default_deps,
+)
+
+gpointerface_a = static_library(
+    'gpointerface_a',
+    'GPOInterface.cpp',
+    dependencies: [default_deps, gpiodcxx],
+)
+
+gpointerface_dep = declare_dependency(
+    include_directories: ['.'],
+    link_with: [gpointerface_a],
     dependencies: default_deps,
 )
 
diff --git a/src/valve-monitor/GPIOValve.cpp b/src/valve-monitor/GPIOValve.cpp
new file mode 100644
index 0000000..f5e501b
--- /dev/null
+++ b/src/valve-monitor/GPIOValve.cpp
@@ -0,0 +1,299 @@
+#include "GPIOValve.hpp"
+
+#include "EntityManagerInterface.hpp"
+#include "GPOInterface.hpp"
+#include "ValveEvents.hpp"
+
+#include <phosphor-logging/lg2.hpp>
+#include <sdbusplus/async.hpp>
+#include <sdbusplus/message/native_types.hpp>
+#include <xyz/openbmc_project/Association/Definitions/aserver.hpp>
+#include <xyz/openbmc_project/Control/Valve/aserver.hpp>
+#include <xyz/openbmc_project/ObjectMapper/client.hpp>
+#include <xyz/openbmc_project/Sensor/Value/aserver.hpp>
+#include <xyz/openbmc_project/State/Decorator/Availability/aserver.hpp>
+#include <xyz/openbmc_project/State/Decorator/OperationalStatus/aserver.hpp>
+
+#include <algorithm>
+#include <array>
+#include <exception>
+#include <functional>
+#include <optional>
+#include <string_view>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+namespace valve
+{
+
+PHOSPHOR_LOG2_USING;
+
+namespace config
+{
+
+/** @brief Polarity name to enum map */
+static constexpr std::array<std::pair<std::string_view, PinPolarity>, 2>
+    validPinPolarity = {
+        {{"Low", PinPolarity::activeLow}, {"High", PinPolarity::activeHigh}}};
+
+auto getConfig(sdbusplus::async::context& ctx,
+               sdbusplus::object_path objectPath)
+    -> sdbusplus::async::task<std::optional<ValveConfig>>
+{
+    ValveConfig config = {};
+    ValveConfigIntf::properties_t properties = {};
+
+    try
+    {
+        properties =
+            co_await ValveConfigIntf(ctx)
+                .service(entity_manager::EntityManagerInterface::serviceName)
+                .path(objectPath.str)
+                .properties();
+    }
+    catch (const std::exception& e)
+    {
+        error("Failed to get properties for {PATH}: {ERR}", "PATH",
+              objectPath.str, "ERR", e);
+        co_return std::nullopt;
+    }
+
+    config.name = properties.name;
+    config.openPinName = properties.open_pin_name;
+
+    for (const auto& [key, value] : config::validPinPolarity)
+    {
+        if (properties.open_polarity == key)
+        {
+            config.openPolarity = value;
+            break;
+        }
+    }
+    if (config.openPolarity == config::PinPolarity::unknown)
+    {
+        error("Invalid polarity {POLARITY} for {NAME}", "POLARITY",
+              properties.open_polarity, "NAME", config.name);
+        co_return std::nullopt;
+    }
+
+    config.openControlPinName = properties.open_control_pin_name;
+    config.openControlValue = properties.open_control_value;
+
+    debug(
+        "Valve config: {NAME} {OPEN_PIN} {OPEN_POLARITY} {OPEN_CONTROL_PIN} {OPEN_CONTROL_VALUE} ",
+        "NAME", config.name, "OPEN_PIN", config.openPinName, "OPEN_POLARITY",
+        config.openPolarity, "OPEN_CONTROL_PIN", config.openControlPinName,
+        "OPEN_CONTROL_VALUE", config.openControlValue);
+
+    co_return config;
+}
+
+} // namespace config
+
+static auto getObjectPath(std::string valveName)
+{
+    std::replace(valveName.begin(), valveName.end(), ' ', '_');
+
+    return (sdbusplus::object_path(ValveIntf::namespace_path::value) /
+            ValveIntf::namespace_path::valve / valveName);
+}
+
+static auto getControlObjectPath(std::string valveName)
+{
+    std::replace(valveName.begin(), valveName.end(), ' ', '_');
+
+    return (sdbusplus::object_path(ValveControlIntf::namespace_path) /
+            valveName);
+}
+
+constexpr ValveIntf::Value::properties_t initValues{0, 100, 0,
+                                                    ValveIntf::Unit::Percent};
+constexpr ValveIntf::Definitions::properties_t initAssociations{};
+constexpr ValveIntf::Availability::properties_t initAvailability{true};
+constexpr ValveIntf::OperationalStatus::properties_t initOperationalState{true};
+constexpr ValveControlIntf::Valve::properties_t initControl{
+    ValveControlIntf::State::Close};
+constexpr ValveControlIntf::Definitions::properties_t initControlAssociations{};
+
+GPIOValve::GPIOValve(sdbusplus::async::context& ctx,
+                     sdbusplus::object_path& objectPath,
+                     Events& events, const config::ValveConfig& config) :
+    ValveIntf(ctx, getObjectPath(config.name).str.c_str(), initValues,
+              initAssociations, initAvailability, initOperationalState),
+    ValveControlIntf(ctx, getControlObjectPath(config.name).str.c_str(),
+                     initControl, initControlAssociations),
+    ctx(ctx),
+    inventoryPath(std::filesystem::path(objectPath.str).parent_path().string()),
+    events(events), config(config),
+    inputInterface(ctx, config.name, config.openPinName,
+                   (config.openPolarity == config::PinPolarity::activeLow),
+                   std::bind_front(&GPIOValve::updateGPIOStateAsync, this))
+{
+    ctx.spawn(inputInterface.start());
+
+    Value::emit_added();
+    OperationalStatus::emit_added();
+    Availability::emit_added();
+    Valve::emit_added();
+
+    info("Created valve {VALVE}", "VALVE", config.name);
+}
+
+GPIOValve::~GPIOValve()
+{
+    available(false);
+}
+
+auto GPIOValve::createAssociations() -> sdbusplus::async::task<>
+{
+    co_await createSensorAssociations();
+
+    createControlAssociations();
+
+    co_return;
+}
+
+auto GPIOValve::get_property(state_t /*unused*/) const -> State
+{
+    debug("Getting {VALVE} state", "VALVE", config.name);
+    return ((value() != 0) ? State::Open : State::Close);
+}
+
+auto GPIOValve::set_property(state_t /*unused*/, auto state) -> bool
+{
+    debug("Setting {VALVE} to {STATE}", "VALVE", config.name, "STATE",
+          convertStateToString(state));
+
+    if ((value() == 0 && state == State::Close) ||
+        (value() != 0 && state == State::Open))
+    {
+        info("Ignoring, as new state {STATE} matches the current state",
+             "STATE", convertStateToString(state));
+        return true;
+    }
+
+    try
+    {
+        gpio::GPOInterface controlGPO{config.name, config.openControlPinName};
+        bool controlValue = ((state == State::Open) ? config.openControlValue
+                                                    : !config.openControlValue);
+        auto res = controlGPO.setValue(controlValue);
+        if (res)
+        {
+            info("Successfully set {VALVE} to {STATE}", "VALVE", config.name,
+                 "STATE", convertStateToString(state));
+            return res;
+        }
+    }
+    catch (const std::exception& e)
+    {
+        error("Failed to set control pin {PIN} to {VALUE}: {ERR}", "PIN",
+              config.openControlPinName, "VALUE", config.openControlValue,
+              "ERR", e);
+    }
+
+    error("Failed to set {VALVE} to {STATE}", "VALVE", config.name, "STATE",
+          convertStateToString(state));
+
+    return false;
+}
+
+auto GPIOValve::updateGPIOStateAsync(bool gpioState) -> sdbusplus::async::task<>
+{
+    auto newValue = gpioState ? 100 : 0;
+
+    if (newValue != value())
+    {
+        debug("Updating valve {VALVE} to {VALUE}", "VALVE", config.name,
+              "VALUE", newValue);
+        value(newValue);
+
+        co_await events.generateValveEvent(inventoryPath, gpioState);
+    }
+
+    co_return;
+}
+
+static auto getContainingChassis(sdbusplus::async::context& ctx,
+                                 std::string& objectPath)
+    -> sdbusplus::async::task<std::optional<std::string>>
+{
+    std::vector<std::string> allInterfaces = {
+        "xyz.openbmc_project.Inventory.Item.Board",
+        "xyz.openbmc_project.Inventory.Item.Chassis",
+    };
+
+    auto client = sdbusplus::client::xyz::openbmc_project::ObjectMapper<>(ctx)
+                      .service("xyz.openbmc_project.ObjectMapper")
+                      .path("/xyz/openbmc_project/object_mapper");
+    auto subTree = co_await client.get_sub_tree(
+        "/xyz/openbmc_project/inventory/system", 2, allInterfaces);
+
+    // A parent that is a chassis takes precedence
+    for (auto& [path, services] : subTree)
+    {
+        if (path == objectPath)
+        {
+            co_return path;
+        }
+    }
+
+    // If no parent is a chassis, return the system chassis
+    for (const auto& [path, services] : subTree)
+    {
+        for (const auto& [service, interfaces] : services)
+        {
+            if (std::find(interfaces.begin(), interfaces.end(),
+                          "xyz.openbmc_project.Inventory.Item.System") !=
+                interfaces.end())
+            {
+                co_return path;
+            }
+        }
+    }
+
+    co_return std::nullopt;
+}
+
+using association_t = std::tuple<std::string, std::string, std::string>;
+using association_list_t = std::vector<association_t>;
+
+auto GPIOValve::createSensorAssociations() -> sdbusplus::async::task<>
+{
+    association_list_t associationList;
+    std::string parent = inventoryPath;
+
+    association_t inventoryAssociation = {"inventory", "sensors",
+                                          inventoryPath};
+
+    associationList.emplace_back(inventoryAssociation);
+
+    auto res = co_await getContainingChassis(ctx, parent);
+
+    association_t chassisAssociation = {"chassis", "all_sensors",
+                                        res.value_or(parent)};
+
+    associationList.emplace_back(chassisAssociation);
+
+    ValveIntf::associations(associationList);
+
+    ValveIntf::Definitions::emit_added();
+
+    co_return;
+}
+
+auto GPIOValve::createControlAssociations() -> void
+{
+    association_list_t associationList;
+
+    association_t association = {"controlling", "controlled_by", inventoryPath};
+
+    associationList.emplace_back(association);
+
+    ValveControlIntf::associations(associationList);
+
+    ValveControlIntf::Definitions::emit_added();
+}
+
+} // namespace valve
diff --git a/src/valve-monitor/GPIOValve.hpp b/src/valve-monitor/GPIOValve.hpp
new file mode 100644
index 0000000..36a692d
--- /dev/null
+++ b/src/valve-monitor/GPIOValve.hpp
@@ -0,0 +1,105 @@
+#pragma once
+
+#include "GPIInterface.hpp"
+#include "GPOInterface.hpp"
+#include "ValveEvents.hpp"
+
+#include <sdbusplus/async.hpp>
+#include <sdbusplus/async/server.hpp>
+#include <sdbusplus/message/native_types.hpp>
+#include <xyz/openbmc_project/Association/Definitions/aserver.hpp>
+#include <xyz/openbmc_project/Configuration/GPIOValve/client.hpp>
+#include <xyz/openbmc_project/Control/Valve/aserver.hpp>
+#include <xyz/openbmc_project/Sensor/Value/aserver.hpp>
+#include <xyz/openbmc_project/State/Decorator/Availability/aserver.hpp>
+#include <xyz/openbmc_project/State/Decorator/OperationalStatus/aserver.hpp>
+
+#include <cstdint>
+#include <optional>
+#include <string>
+
+namespace valve
+{
+
+using ValveConfigIntf =
+    sdbusplus::client::xyz::openbmc_project::configuration::GPIOValve<>;
+
+class GPIOValve;
+
+using ValveIntf = sdbusplus::async::server_t<
+    GPIOValve, sdbusplus::aserver::xyz::openbmc_project::sensor::Value,
+    sdbusplus::aserver::xyz::openbmc_project::association::Definitions,
+    sdbusplus::aserver::xyz::openbmc_project::state::decorator::Availability,
+    sdbusplus::aserver::xyz::openbmc_project::state::decorator::
+        OperationalStatus>;
+
+using ValveControlIntf = sdbusplus::async::server_t<
+    GPIOValve, sdbusplus::aserver::xyz::openbmc_project::control::Valve,
+    sdbusplus::aserver::xyz::openbmc_project::association::Definitions>;
+
+namespace config
+{
+
+enum class PinPolarity
+{
+    activeLow,
+    activeHigh,
+    unknown
+};
+
+struct ValveConfig
+{
+    std::string name = Defaults::name;
+    std::string openPinName = Defaults::pinName;
+    PinPolarity openPolarity = PinPolarity::unknown;
+    std::string openControlPinName = Defaults::pinName;
+    bool openControlValue = false;
+    uint8_t targetOpen = 100; // For a GPIO valve, 100% is the target open
+                              // value as valve can only be open or closed.
+
+    struct Defaults
+    {
+        static constexpr auto name = "unknown";
+        static constexpr auto pinName = "unknown";
+    };
+};
+
+/** @brief Get the valve configuration from the Entity Manager */
+auto getConfig(sdbusplus::async::context& ctx,
+               sdbusplus::object_path objectPath)
+    -> sdbusplus::async::task<std::optional<ValveConfig>>;
+
+} // namespace config
+
+class GPIOValve : public ValveIntf, public ValveControlIntf
+{
+  public:
+    explicit GPIOValve(sdbusplus::async::context& ctx,
+                       sdbusplus::object_path& objectPath,
+                       Events& events, const config::ValveConfig& config);
+
+    ~GPIOValve();
+
+    auto createAssociations() -> sdbusplus::async::task<>;
+
+    // NOLINTNEXTLINE(readability-identifier-naming)
+    auto get_property(state_t /*unused*/) const -> State;
+
+    // NOLINTNEXTLINE(readability-identifier-naming)
+    auto set_property(state_t /*unused*/, auto state) -> bool;
+
+  private:
+    auto updateGPIOStateAsync(bool gpioState) -> sdbusplus::async::task<>;
+
+    auto createSensorAssociations() -> sdbusplus::async::task<>;
+
+    auto createControlAssociations() -> void;
+
+    sdbusplus::async::context& ctx;
+    sdbusplus::object_path inventoryPath;
+    Events& events;
+    config::ValveConfig config;
+    gpio::GPIInterface inputInterface;
+};
+
+} // namespace valve
diff --git a/src/valve-monitor/Readme.md b/src/valve-monitor/Readme.md
new file mode 100644
index 0000000..fb849e5
--- /dev/null
+++ b/src/valve-monitor/Readme.md
@@ -0,0 +1,50 @@
+# Valve Monitor
+
+The valve monitor consumes the EM Configuration and provides the open status of
+the specified valve as sensor data. It also generates an event message whenever
+the valve's open status changes. Additionally, it offers a control interface to
+open and close the valve.
+
+## EM Configuration
+
+The example of EM Configuration is as follows:
+
+```json
+{
+  "Name": "ReturnValve",
+  "OpenControlPinName": "WATER_VALVE_CLOSED_N",
+  "OpenControlValue": false,
+  "OpenPinName": "RETURN_CNTL_FB_D_R",
+  "OpenPolarity": "High",
+  "Type": "GPIOValve"
+}
+```
+
+For the GPIO hardware interface, the valve's open status is indicated as a
+percentage, with 100% representing fully open and 0% representing fully closed,
+based on the GPIO signal state and its polarity.
+
+## D-Bus Interfaces
+
+The following D-Bus interfaces are implemented:
+
+- xyz.openbmc_project.Sensor.Value
+- xyz.openbmc_project.State.Decorator.OperationalStatus
+- xyz.openbmc_project.State.Decorator.Availability
+- xyz.openbmc_project.Association.Definitions
+
+  Above interfaces are exposed on the following object path:
+
+  /xyz/openbmc_project/sensors/valve/\<ValveName\>
+
+  where \<ValveName\> corresponds to the valve name specified in the EM
+  Configuration.
+
+- xyz.openbmc_project.Control.Valve
+
+  Above interface is exposed on the following object path:
+
+  /xyz/openbmc_project/control/valve/\<ValveName\>
+
+  where \<ValveName\> corresponds to the valve name specified in the EM
+  Configuration.
diff --git a/src/valve-monitor/ValveEvents.cpp b/src/valve-monitor/ValveEvents.cpp
new file mode 100644
index 0000000..b965c86
--- /dev/null
+++ b/src/valve-monitor/ValveEvents.cpp
@@ -0,0 +1,34 @@
+#include "ValveEvents.hpp"
+
+#include <phosphor-logging/commit.hpp>
+#include <phosphor-logging/lg2.hpp>
+#include <sdbusplus/async.hpp>
+#include <sdbusplus/message/native_types.hpp>
+#include <xyz/openbmc_project/State/Valve/event.hpp>
+
+namespace valve
+{
+
+PHOSPHOR_LOG2_USING;
+
+auto Events::generateValveEvent(sdbusplus::object_path valvePath,
+                                bool open) -> sdbusplus::async::task<>
+{
+    namespace event_intf = sdbusplus::event::xyz::openbmc_project::state::Valve;
+
+    if (open)
+    {
+        co_await lg2::commit(ctx,
+                             event_intf::ValveOpen("VALVE_NAME", valvePath));
+    }
+    else
+    {
+        co_await lg2::commit(ctx,
+                             event_intf::ValveClose("VALVE_NAME", valvePath));
+    }
+
+    debug("Valve event generated for {PATH} with open={STATUS}", "PATH",
+          valvePath, "STATUS", open);
+}
+
+} // namespace valve
diff --git a/src/valve-monitor/ValveEvents.hpp b/src/valve-monitor/ValveEvents.hpp
new file mode 100644
index 0000000..b6fc054
--- /dev/null
+++ b/src/valve-monitor/ValveEvents.hpp
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <sdbusplus/async.hpp>
+#include <sdbusplus/message/native_types.hpp>
+
+namespace valve
+{
+
+class Events
+{
+  public:
+    Events() = delete;
+
+    explicit Events(sdbusplus::async::context& ctx) : ctx(ctx) {}
+
+    auto generateValveEvent(sdbusplus::object_path valvePath,
+                            bool open) -> sdbusplus::async::task<>;
+
+  private:
+    sdbusplus::async::context& ctx;
+};
+
+} // namespace valve
diff --git a/src/valve-monitor/ValveMonitor.cpp b/src/valve-monitor/ValveMonitor.cpp
new file mode 100644
index 0000000..397e86b
--- /dev/null
+++ b/src/valve-monitor/ValveMonitor.cpp
@@ -0,0 +1,102 @@
+#include "ValveMonitor.hpp"
+
+#include "GPIOValve.hpp"
+
+#include <phosphor-logging/lg2.hpp>
+#include <sdbusplus/async.hpp>
+#include <sdbusplus/message/native_types.hpp>
+#include <sdbusplus/server/manager.hpp>
+
+#include <exception>
+#include <functional>
+#include <memory>
+#include <string>
+#include <utility>
+
+PHOSPHOR_LOG2_USING;
+
+namespace valve
+{
+
+ValveMonitor::ValveMonitor(sdbusplus::async::context& ctx) :
+    ctx(ctx), events(ctx),
+    entityManager(ctx, {ValveConfigIntf::interface},
+                  std::bind_front(&ValveMonitor::processInventoryAdded, this),
+                  std::bind_front(&ValveMonitor::processInventoryRemoved, this))
+{
+    ctx.spawn(entityManager.handleInventoryGet());
+}
+
+auto ValveMonitor::processInventoryAdded(
+    const sdbusplus::object_path& objectPath,
+    const std::string& /*unused*/) -> void
+{
+    // Added NO_LINT to bypass clang-tidy warning about STDEXEC_ASSERT as clang
+    // seems to be confused about context being uninitialized.
+    // NOLINTNEXTLINE(clang-analyzer-core.uninitialized.Branch)
+    ctx.spawn(processConfigAddedAsync(objectPath));
+}
+
+auto ValveMonitor::processInventoryRemoved(
+    const sdbusplus::object_path& objectPath,
+    const std::string& /*unused*/) -> void
+{
+    if (!valves.contains(objectPath.str))
+    {
+        return;
+    }
+    debug("Removed valve {VALVE}", "VALVE", objectPath);
+    valves.erase(objectPath.str);
+}
+
+auto ValveMonitor::processConfigAddedAsync(
+    sdbusplus::object_path objectPath) -> sdbusplus::async::task<>
+{
+    auto res = co_await config::getConfig(ctx, objectPath);
+    if (!res.has_value())
+    {
+        error("Failed to get config for {OBJECT_PATH}", "OBJECT_PATH",
+              objectPath);
+        co_return;
+    }
+    auto config = res.value();
+
+    if (valves.contains(objectPath.str))
+    {
+        warning("Valve {VALVE} already exist", "VALVE", config.name);
+        co_return;
+    }
+
+    try
+    {
+        auto valve =
+            std::make_unique<GPIOValve>(ctx, objectPath, events, config);
+        co_await valve->createAssociations();
+        valves[objectPath.str] = std::move(valve);
+    }
+    catch (std::exception& e)
+    {
+        error("Failed to create valve {VALVE}: {ERROR}", "VALVE", config.name,
+              "ERROR", e);
+    }
+
+    co_return;
+}
+
+} // namespace valve
+
+int main()
+{
+    constexpr auto path = "/xyz/openbmc_project";
+    constexpr auto serviceName = "xyz.openbmc_project.valvemonitor";
+    sdbusplus::async::context ctx;
+    sdbusplus::server::manager_t manager{ctx, path};
+
+    info("Creating valve monitor at {PATH}", "PATH", path);
+    valve::ValveMonitor valveMonitor{ctx};
+
+    ctx.request_name(serviceName);
+
+    ctx.run();
+    return 0;
+}
diff --git a/src/valve-monitor/ValveMonitor.hpp b/src/valve-monitor/ValveMonitor.hpp
new file mode 100644
index 0000000..fe6c733
--- /dev/null
+++ b/src/valve-monitor/ValveMonitor.hpp
@@ -0,0 +1,49 @@
+#include "EntityManagerInterface.hpp"
+#include "GPIOValve.hpp"
+#include "ValveEvents.hpp"
+
+#include <sdbusplus/async.hpp>
+#include <sdbusplus/message/native_types.hpp>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+namespace valve
+{
+
+class ValveMonitor
+{
+  public:
+    ValveMonitor() = delete;
+    ValveMonitor(const ValveMonitor&) = delete;
+    ValveMonitor(ValveMonitor&&) = delete;
+    ValveMonitor& operator=(const ValveMonitor&) = delete;
+    ValveMonitor& operator=(ValveMonitor&&) = delete;
+
+    explicit ValveMonitor(sdbusplus::async::context& ctx);
+
+  private:
+    using valve_map_t =
+        std::unordered_map<std::string, std::unique_ptr<GPIOValve>>;
+
+    /** @brief  Process new interfaces added to inventory */
+    auto processInventoryAdded(
+        const sdbusplus::object_path& objectPath,
+        const std::string& interfaceName) -> void;
+
+    /** @brief Process interfaces removed from inventory */
+    auto processInventoryRemoved(
+        const sdbusplus::object_path& objectPath,
+        const std::string& interfaceName) -> void;
+
+    /** @brief Process the config add asynchronously */
+    auto processConfigAddedAsync(sdbusplus::object_path objectPath)
+        -> sdbusplus::async::task<>;
+
+    sdbusplus::async::context& ctx;
+    Events events;
+    entity_manager::EntityManagerInterface entityManager;
+    valve_map_t valves;
+};
+} // namespace valve
diff --git a/src/valve-monitor/meson.build b/src/valve-monitor/meson.build
index d881ba8..d00cefa 100644
--- a/src/valve-monitor/meson.build
+++ b/src/valve-monitor/meson.build
@@ -1,3 +1,25 @@
 if not get_option('valve-monitor').allowed()
     subdir_done()
 endif
+
+src_inc = include_directories('..')
+phosphor_dbus_interfaces_dep = dependency('phosphor-dbus-interfaces')
+
+executable(
+    'valvemonitor',
+    'ValveMonitor.cpp',
+    'GPIOValve.cpp',
+    'ValveEvents.cpp',
+    dependencies: [
+        default_deps,
+        utils_dep,
+        phosphor_dbus_interfaces_dep,
+        gpiodcxx,
+        gpiinterface_dep,
+        gpointerface_dep,
+        entitymanagerinterface_dep,
+    ],
+    include_directories: src_inc,
+    install: true,
+    install_dir: '/usr/libexec/dbus-sensors',
+)
-- 
2.53.0-Meta

